### PR TITLE
archive: CArchiveWriter: strip the path from .pyc modules

### LINKED ISSUE #5377: Build not reproducible when different python environement used
I'm not sure whether this is a bug but I found that a build on Windows is not reproducible when a different virtualenv is used even when both environements have exactly the same packages.

Im using 
Python 3.8.3
pyinstaller 4.1
Windows 10

What I did:
open cmd
mkdir folder_1
cd folder_1
echo print('hello world') >> test.py
set PYTHONHASHSEED=1
C:\path_to_python_38\python.exe -m virtualenv venv
venv\Scripts\activate
python -m pip install pyinstaller
python -m PyInstaller --clean test.py

:: md5 checksum data of: dist\test				       16C651921E1AF9A6CC5B539E4AE772C0
:: md5 checksum data and names of: dist\test	 	4521166FB1B3B049CB7FEEF23FAD181E
:: md5 checksum data of: dist\test\test.exe	 	       3A2E7CDB2FF328172FEF5ECC4805E99B

open cmd
mkdir folder_2
cd folder_2
echo print('hello world') >> test.py
set PYTHONHASHSEED=1
C:\path_to_python_38\python.exe -m virtualenv venv
venv\Scripts\activate
python -m pip install pyinstaller
python -m PyInstaller --clean test.py

:: md5 checksum data of: dist\test				        7FB0108BEB9E484B99D935A1D8121934
:: md5 checksum data and names of: dist\test	 	C8A76D3B688AD89536ABC1A0A5841482
:: md5 checksum data of: dist\test\test.exe	 	        A3183BD4FC7778BBFB6C41CFD6308F0F

This implies that a third person is probably not able to reproduce a given binary and even on the same computer when a folder is changed, the binary will differ. I'm not sure wether this is even a desired feature and maybe I understand this wrong, but what is the point of reproducibility if you cannot reproduce the binary file on a third machine in order to check wether a given binary actually is the one corresponding to a given code.

### PR DESCRIPTION
**Pull request by rokm**
This makes the .pyc modules in the PKG archive invariant to the build-environment location, and in turn enables reproducible
builds without having to match the location of the build environment.

Fixes #5377.

### CODE REVIEW

File: PyInstaller/archive/writers.py [hunk 1]
```diff
@@ -375,6 +376,26 @@ def add(self, entry):
...
+                ulen = os.fstat(fh.fileno()).st_size
+                # Check if it is a PYC file
+                header = fh.read(4)
+                fh.seek(0)
+                if header == BYTECODE_MAGIC:
+                    # Read whole header and load code
+                    if is_py37:
+                        header = fh.read(16)
+                    else:
+                        header = fh.read(12)
```
Comment (bwoodsend) [inline PyInstaller/archive/writers.py:390]: Do you have a link you can comment in to say where the 16 and the 12 come from?

Comment (rokm) [reply PyInstaller/archive/writers.py:397]: I've added a comment about header sizes and reference to PEP-552. 

And also a comment on why original header can be re-used after we modify the code object.

Comment (bwoodsend) [reply PyInstaller/archive/writers.py:397]: Perfect.

**Review by bwoodsend — APPROVED** [review]

**Comment by Legorooj** [comment]
Nice work!

**Comment by htgoebel** [comment]
Does this also solve #2494?

**Comment by rokm** [comment]
> Does this also solve #2494?

I think it should, yes, as the leaked path in that issue is coming from `pyimod03_importers.py`.

**Pull request merged**